Mathlib/LinearAlgebra/Matrix/NonsingularInverse.lean

```diff
@@ -361,6 +368,87 @@ lemma mul_right_injective_of_inv (A : Matrix m n α) (B : Matrix n m α) (h : A
 
 end InjectiveMul
 
+section vecMul
+
+variable [Fintype m] [DecidableEq m] {K : Type*} [Field K]
+
+
+theorem vecMul_surjective_iff_exists_left_inverse {A : Matrix m n α} :
+    Function.Surjective A.vecMul ↔ ∃ B : Matrix n m α, B * A = 1 := by
+  refine ⟨fun h ↦ ?_, fun ⟨B, hBA⟩ y ↦ ⟨B.vecMul y, by simp [hBA]⟩⟩
+  choose rows hrows using (h <| Pi.single · 1)
+  refine ⟨Matrix.of rows, Matrix.ext fun i j => ?_⟩
+  rw [mul_apply_eq_vecMul, one_eq_pi_single, ←hrows]
+  rfl
+
+theorem mulVec_surjective_iff_exists_right_inverse {A : Matrix m n α} :
+    Function.Surjective A.mulVec ↔ ∃ B : Matrix n m α, A * B = 1 := by
+  conv_lhs => arg 1; eta_expand
+  simp_rw [← vecMul_transpose, vecMul_surjective_iff_exists_left_inverse]
+  refine ⟨fun ⟨B, hB⟩ ↦ ⟨Bᵀ, ?_⟩, fun ⟨B, hB⟩ ↦ ⟨Bᵀ, by rw [←transpose_mul, hB, transpose_one]⟩⟩
+  apply_fun Matrix.transpose using transpose_injective
+  rw [transpose_mul, transpose_transpose, hB, transpose_one]
+
+theorem vecMul_surjective_iff_isUnit {A : Matrix m m α} :
+    Function.Surjective A.vecMul ↔ IsUnit A := by
+  rw [vecMul_surjective_iff_exists_left_inverse, exists_left_inverse_iff_isUnit]
+
+theorem mulVec_surjective_iff_isUnit {A : Matrix m m α} :
+    Function.Surjective A.mulVec ↔ IsUnit A := by
+  rw [mulVec_surjective_iff_exists_right_inverse, exists_right_inverse_iff_isUnit]
+
+theorem vecMul_injective_iff_isUnit {A : Matrix m m K} :
+    Function.Injective A.vecMul ↔ IsUnit A := by
+  refine ⟨fun h ↦ ?_, fun h ↦ ?_⟩
+  · rw [← vecMul_surjective_iff_isUnit]
+    exact LinearMap.surjective_of_injective (f := A.vecMulLinear) h
+  change Function.Injective A.vecMulLinear
+  rw [← LinearMap.ker_eq_bot, LinearMap.ker_eq_bot']
+  intro c hc
+  replace h := h.invertible
+  simpa using congr_arg A⁻¹.vecMulLinear hc
```

Comment:
I think this one should hold for division rings? Note that since you have `h`, you don't need `A⁻¹`, and can use `invOf` (`\frac1`) instead, which removes the commutativity assumption.

Reply:
I'm not sure about the forwards direction - you can get to the existence of a left inverse from injectivity, but the lemma that gets you from there to `IsUnit A` requires `CommRing`. Is it true that left-invertibility implies right-invertibility for matrices over noncommutative rings?

Reply:
Ah you're right, of course it doesn't; you need only consider 1x1 matrices to see why.

Reply:
Perhaps this is evidence that these should be a separate lemma for existence of a left-inverse.

Reply:
Perhaps - this feels like scope creep for this PR, though.

Reply:
For me, the fact that unneeded `Fintype`s are scattered around the matrix API is more of a pressing problem than the commutativity assumptions. But that's for another day as well.